engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/bool/AndNode.java

```diff
@@ -1,14 +1,76 @@
 package org.enso.interpreter.node.expression.builtin.bool;
 
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
 import com.oracle.truffle.api.nodes.Node;
 import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.dsl.MonadicState;
+import org.enso.interpreter.dsl.Suspend;
+import org.enso.interpreter.node.BaseNode;
+import org.enso.interpreter.node.callable.thunk.ThunkExecutorNode;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.state.Stateful;
 
 @BuiltinMethod(
     type = "Boolean",
     name = "&&",
     description = "Computes the logical conjunction of two booleans")
-public class AndNode extends Node {
-  boolean execute(boolean _this, boolean that) {
-    return _this && that;
+public abstract class AndNode extends Node {
+
+  public static AndNode build() {
+    return AndNodeGen.create();
+  }
+
+  abstract Stateful execute(@MonadicState Object state, boolean _this, @Suspend Object that);
+
+  @Specialization(guards = {"!_this"})
+  Stateful executeLhs(@MonadicState Object state, boolean _this, @Suspend Object that) {
+    return new Stateful(state, false);
+  }
+
+  @Specialization(
+      guards = {"_this"},
+      rewriteOn = ClassCastException.class)
+  Stateful executeBool(
+      @MonadicState Object state,
+      boolean _this,
+      @Suspend Object that,
+      @Cached ThunkExecutorNode rhsThunkExecutorNode) {
+    Stateful result =
+        rhsThunkExecutorNode.executeThunk(that, state, BaseNode.TailStatus.TAIL_DIRECT);
+    if (result.getValue() instanceof Boolean) {
+      return result;
+    }
+    throw new ClassCastException("expected Boolean");
+  }
+
+  @Specialization(
+      guards = {"_this"},
+      rewriteOn = ClassCastException.class)
+  Stateful executeDataflowError(
+      @MonadicState Object state,
+      boolean _this,
+      @Suspend Object that,
+      @Cached ThunkExecutorNode rhsThunkExecutorNode) {
+    Stateful result =
+        rhsThunkExecutorNode.executeThunk(that, state, BaseNode.TailStatus.TAIL_DIRECT);
+    if (result.getValue() instanceof DataflowError) {
+      return result;
+    }
+    throw new ClassCastException("expected DataflowError");
```

Comment:
Using `@Specialization` makes the code more Truffle-like. However I am still not able to estimate the impact of these changes on IGV graphs. I tried to [write a LogicalBenchmarks suite](https://github.com/enso-org/enso/commit/7f9113ea6ce7d7eb7292998551453314eca65145) and check the graphs. I can then launch IGV and
```
sbt:runtime> withDebug --dumpGraphs benchOnly -- LogicalBenchmarks
```
and obtain a bunch of graphs. 
```

Reply:
I modified the way dataflowerrors are handled. Seemed like I even managed to get a consistent speedup in that benchmark